### DISCUSSION TITLE
Usage of IDbContextFactory

### DISCUSSION THREAD

**Discussion by advapiIT** [linq2db/linq2db#3928] [Q&A]
Hello,
I've used in the past Linq2Db, now on an existing application that make large use of EF Core 6 , I wish to use Linq2DB to perform some strategic queries that take long time when using EF classes.. 

I looked at this [thread](https://github.com/linq2db/linq2db/issues/390) since I wish to open multiple connection to retrieve data in parallel (I use repositories, want to use Task.WhenAll to fill portions of my DTO).

I was wondering using IDbContextFactory will use the ADO.NET ConnectionPool as well? anyone can share his/her implementation?

Thanks

**Comment by jods4** [comment]
Most ADO.NET providers use a connection pool to improve their performance.
This happens at the ADO.NET level, i.e. `SqlConnection`, `OracleConnection`, etc.

Linq2db is a higher level library, built on top of ADO.NET. It has to use ADO.NET `DbConnection`s to connect to your database, as any other code your write would.
Thus, connection pooling from your DB provider will happen as usual.

Or if your question was: are linq2db's `DbContext` or `DataConnection` pooled? the answer is no.
I suppose you could pool them if you want in an `ObjectPool`, but be mindful of opening/closing the underlying ADO.NET `DbConnection` in an efficient manner, so that your object pool doesn't keep connections alive and starve the underlying ADO.NET connection pool.

Keep in mind that ADO.NET `DbConnection`s actually establish network connections and authentication with other servers and this is a costly operation. That's the main reason why connections are kept open and pooled for re-use.
`DbContext` is just a .net object that sits on top of a connection, so unless you really want to reduce GC allocations it makes little sense to pool it.

**Comment by advapiIT** [comment]
Hello @jods4 , thanks for your fast reply,
so you say that it's best for me just to inject the DbContext and use it for all the operations instead of having each method to create a new connection from DbFactory item?

**Reply by jods4** [reply]
I don't know the context of your application, so take this with a grain of salt:

Assuming it's a web application, it would be reasonable to have `DbContext` registered as a _Scoped_ component in your DI.
So you can inject and use a single `DbContext` (i.e. a single db connection) per request, and have it automatically disposed when the request ends.

It would also be a good idea to register a `IDbContextFactory` (or even just a `Func<DbContext>` if your IoC supports it) as a _Singleton_, so that you can create additional `DbContext`s (that you must dispose yourself) when you need to open multiple connections to perform operations in parallel.

Remember that `DbContext` is a wrapper around an ADO.NET connection, so it's not thread-safe and should not be shared.
